File: 123-best-time-to-buy-and-sell-stock-iii/best-time-to-buy-and-sell-stock-iii.cpp
class Solution {
public: 
      int find(int i,int buy,int k,int n,vector<int>&p,vector<vector<vector<int>>>&v){
        if(i==n || k==0) return 0;
        if(v[i][buy][k]!=-1) return v[i][buy][k];

        if(buy){
            v[i][buy][k]=max(-p[i]+find(i+1,0,k,n,p,v),0+find(i+1,1,k,n,p,v));
        }else{
            v[i][buy][k]=max(p[i]+find(i+1,1,k-1,n,p,v),0+find(i+1,0,k,n,p,v));
        }
        return v[i][buy][k];
    }
    int maxProfit(vector<int>& p) {
         int i=0,n=p.size(),k=2;
        vector<vector<vector<int>>> v(n+1,vector<vector<int>>(2, vector<int>(k + 1, -1)));
         return find(0,1,k,n,p,v);
    }
};

File: 123-best-time-to-buy-and-sell-stock-iii/best-time-to-buy-and-sell-stock-iii.java
import java.util.Arrays;

class Solution {

    int solve(int[] prices, int buy,
              int i, int[][][] dp,
              int k) {

        if(i == prices.length || k == 0)
            return 0;

        if(dp[i][buy][k] != -1)
            return dp[i][buy][k];

        int ans;

        if(buy == 1) {

            ans = Math.max(
                    -prices[i] + solve(prices, 0, i + 1, dp, k),
                    solve(prices, 1, i + 1, dp, k)
                  );

        } else {

            ans = Math.max(
                    prices[i] + solve(prices, 1, i + 1, dp, k - 1),
                    solve(prices, 0, i + 1, dp, k)
                  );
        }

        return dp[i][buy][k] = ans;
    }

    public int maxProfit(int[] prices) {

        int n = prices.length;
        int k = 2;

        int[][][] dp = new int[n][2][k + 1];

        for(int i = 0; i < n; i++) {
            for(int j = 0; j < 2; j++) {
                Arrays.fill(dp[i][j], -1);
            }
        }

        return solve(prices, 1, 0, dp, k);
    }
}